File: Commons/Action/src/main/java/com/greentree/commons/action/observable/RunObservable.java
package com.greentree.commons.action.observable;

import com.greentree.commons.action.ListenerCloser;

import java.io.Serializable;

public interface RunObservable extends Serializable {

    RunObservable NULL = new RunObservable() {

        private static final long serialVersionUID = 1L;

        @Override
        public ListenerCloser addListener(Runnable listener) {
            return ListenerCloser.EMPTY;
        }

        @Override
        public int listenerSize() {
            return 0;
        }
    };

    ListenerCloser addListener(Runnable listener);

    default boolean isEmpty() {
        return listenerSize() == 0;
    }

    default int listenerSize() {
        throw new UnsupportedOperationException();
    }

}

File: Commons/Action/src/main/java/com/greentree/commons/action/observable/RunObservable.kt
package com.greentree.commons.action.observable

import com.greentree.commons.action.ListenerCloser
import java.io.Serializable

interface RunObservable : AutoCloseable, Serializable {

	fun addListener(listener: Runnable): ListenerCloser

	val isEmpty: Boolean
		get() = listenerCount == 0
	val listenerCount: Int
		get() = TODO()

	data object Empty : RunObservable {

		override fun addListener(listener: Runnable): ListenerCloser {
			return ListenerCloser.Empty
		}

		override val listenerCount: Int
			get() = 0

		override fun close() {
		}
	}

	companion object {

		@Deprecated("use object", ReplaceWith("RunObservable.Empty"))
		@JvmField
		val NULL = Empty
	}
}
